### PR TITLE
MM-53924 - Implement push notifications plugin hook and plugin api method

### PR DESCRIPTION
**Pull request by cpoile**
#### Summary
- Based on discussions with @streamer45 and @lieut-data (in previous PRs), agree that the more powerful solution is probably best: put the hook and the plugin-api call closer to where the notifications are sent.
- I've added comments to the plugin-api method to warn of the dangers of this. It's up to the plugins to properly assemble the push notifications.
- Reverts #24263 and #24273 and reimplements them.
- Closes #24333 

cc @amyblais Similar to #24333 I think this is safe for 9.0 (no dependencies are affected)

#### Ticket Link
- https://mattermost.atlassian.net/browse/MM-53924

#### Release Note
```release-note
Added a NotificationWillBePushed plugin hook invoked before the push notification is processed and sent to the notification service. Plugins may modify or reject the push notification.
Added a `SendPushNotification` plugin api method which allows plugins to send push notifications to a specific user's mobile sessions.
```

### CODE REVIEW

**Review by streamer45 — APPROVED** [review]
Nice work :+1:

File: server/channels/app/notification_push.go [hunk 1]
```diff
@@ -77,10 +77,33 @@ func (a *App) sendPushNotificationSync(c request.CTX, post *model.Post, user *mo
 		return appErr
 	}
 
+	// Add post and channel types for plugins to use in the NotificationWillBePushed hook
+	msg.PostType = post.Type
+	msg.ChannelType = channel.Type
```
[the remark this answers is not in the document]
Comment (cpoile) [reply server/channels/app/notification_push.go]: Yep, looks better there, thanks!

File: server/public/plugin/api.go [hunk 2]
```diff
@@ -1181,14 +1181,17 @@ type API interface {
 	// Minimum server version: 7.6
 	GetUploadSession(uploadID string) (*model.UploadSession, error)
 
-	// SendPluginPushNotification will attempt to send a push notification to `notification.User`, using
-	// `notification.Post` as the source of the notification. The server will use the PluginPushNotification
-	// data to construct the final push notification according to the server's configuration and license. Refer
-	// to `App.BuildPushNotificationMessage` for the logic used to construct the push notification.
-	// Note: the NotificationWillBePushed hook will be run after SendPluginPushNotification is called.
+	// SendPushNotification will send a push notification to all of user's sessions, using
+	//
```
Comment (isacikgoz) [inline server/public/plugin/api.go:1185]: Should these lines removed?

**Review by isacikgoz — APPROVED** [review]
LGTM, just a note an a comment.

File: server/public/plugin/api.go (hunk 2 above)
Comment (cpoile) [reply server/public/plugin/api.go]: Ah, slight mistake, thanks. Fixed!

**Pull request merged**